### PR TITLE
Powershell support for IpGroup Resource

### PR DESCRIPTION
**Pull request by ssripadham**
## Description

We are building a facility in Azure Networking that will allow Azure customers to group of Ip Address prefixes to define rules in Azure Networking. This will help customers manage & minimize the overall complexity for security rules creation.
Customer will create an IpGroup (new top-level resource) and specify a set of IP addresses to include within the tag. These tags will then be used by existing resources as Azure Firewall. 

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by wyunchi-ms** [comment]
Here is the design issue https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/411

File: src/Network/Network.Test/ScenarioTests/IpGroupTest.ps1 [hunk 1]
```diff
@@ -0,0 +1,70 @@
+﻿# ----------------------------------------------------------------------------------
```
Comment (number213) [inline src/Network/Network.Test/ScenarioTests/IpGroupTest.ps1]: Extra file? There's already `IpGroupsTests.ps1`

**Comment by number213** [comment]
Please also resolve conflicts and check errors

File: src/Network/Network.Test/ScenarioTests/IpGroupTest.ps1 (hunk 1 above)
Comment (ssripadham) [reply src/Network/Network.Test/ScenarioTests/IpGroupTest.ps1]: I have removed IpGroupsTests.ps1

Comment (number213) [reply src/Network/Network.Test/ScenarioTests/IpGroupTest.ps1]: ps1 file should have the same name cs file does, so your test files should be called `IpGroupTests.ps1`

File: src/Network/Network/IpGroup/GetIpGroupCommand.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    public class GetIpGroupsCommand : IpGroupBaseCmdlet
+    {
+        [Parameter(
+            ParameterSetName = IpGroupParameterSetNames.ByName,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The resource group name.")]
+        [ValidateNotNullOrEmpty]
+        [SupportsWildcards]
+        public virtual string ResourceGroupName { get; set; }
```
Comment (number213) [inline src/Network/Network/IpGroup/GetIpGroupCommand.cs]: Please add `[ResourceGroupCompleter]`

File: src/Network/Network/IpGroup/NewIpGroupCommand.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+         HelpMessage = "The resource name.")]
+        [ValidateNotNullOrEmpty]
+        public virtual string Name { get; set; }
+
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The resource group name.")]
+        [ValidateNotNullOrEmpty]
+        public virtual string ResourceGroupName { get; set; }
```
Comment (number213) [inline src/Network/Network/IpGroup/NewIpGroupCommand.cs]: Please add `[ResourceGroupCompleter]`

File: src/Network/Network/IpGroup/NewIpGroupCommand.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            HelpMessage = "IpAddresses defined in the IpGroup")]
+        [ValidateNotNullOrEmpty]
+        public string[] IpAddress { get; set; }
+
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "location.")]
+        [ValidateNotNullOrEmpty]
+        public virtual string Location { get; set; }
```
Comment (number213) [inline src/Network/Network/IpGroup/NewIpGroupCommand.cs]: Please add `[LocationCompleter("Microsoft.Network/ipGroups")]`

File: src/Network/Network/IpGroup/RemoveIpGroupCommand.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        [ValidateNotNullOrEmpty]
+        public virtual string Name { get; set; }
+
+        [Parameter(
+            ParameterSetName = IpGroupParameterSetNames.ByName,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The resource group name.")]
+        [ValidateNotNullOrEmpty]
+        public virtual string ResourceGroupName { get; set; }
```
Comment (number213) [inline src/Network/Network/IpGroup/RemoveIpGroupCommand.cs]: Please add `[ResourceGroupCompleter]`

File: src/Network/Network/IpGroup/GetIpGroupCommand.cs [hunk 6]
```diff
@@ -34,6 +34,7 @@ public class GetIpGroupsCommand : IpGroupBaseCmdlet
             HelpMessage = "The resource group name.")]
         [ValidateNotNullOrEmpty]
         [SupportsWildcards]
+        [ResourceNameCompleter("Microsoft.Network/ipGroups", "ResourceGroupName")]
```
Comment (number213) [inline src/Network/Network/IpGroup/GetIpGroupCommand.cs]: It should be `[ResourceGroupCompleter]`

File: src/Network/Network/IpGroup/NewIpGroupCommand.cs [hunk 7]
```diff
@@ -43,6 +44,7 @@ public class NewIpGroupsCommand : IpGroupBaseCmdlet
             ValueFromPipelineByPropertyName = true,
             HelpMessage = "The resource group name.")]
         [ValidateNotNullOrEmpty]
+        [ResourceNameCompleter("Microsoft.Network/ipGroups", "ResourceGroupName")]
```
Comment (number213) [inline src/Network/Network/IpGroup/NewIpGroupCommand.cs]: It should be `[ResourceGroupCompleter]`

File: src/Network/Network/IpGroup/RemoveIpGroupCommand.cs [hunk 8]
```diff
@@ -40,6 +40,7 @@ public class RemoveIpGroupsCommand : IpGroupBaseCmdlet
             ValueFromPipelineByPropertyName = true,
             HelpMessage = "The resource group name.")]
         [ValidateNotNullOrEmpty]
+        [ResourceNameCompleter("Microsoft.Network/ipGroups", "ResourceGroupName")]
```
Comment (number213) [inline src/Network/Network/IpGroup/RemoveIpGroupCommand.cs]: It should be `[ResourceGroupCompleter]`

File: src/Network/Network.Test/ScenarioTests/IpGroupTest.ps1 (hunk 1 above)
Comment (ssripadham) [reply src/Network/Network.Test/ScenarioTests/IpGroupTest.ps1]: This file has been removed

File: src/Network/Network/IpGroup/NewIpGroupCommand.cs (hunk 7 above)
Comment (ssripadham) [reply src/Network/Network/IpGroup/NewIpGroupCommand.cs]: Done

File: src/Network/Network/IpGroup/GetIpGroupCommand.cs (hunk 6 above)
Comment (ssripadham) [reply src/Network/Network/IpGroup/GetIpGroupCommand.cs]: Dne

**Comment by VeryEarly** [comment]
@ssripadham

below error was thrown during help generation, please try
"msbuild build.proj /t:GenerateHelp "
at your local to find and fix the error.

```text
[command]"C:\Program Files\dotnet\dotnet.exe" msbuild build.proj /t:GenerateHelp /p:Configuration=Debug;PullRequestNumber=$(System.PullRequest.PullRequestNumber)
2019-10-26T01:01:32.5413691Z 
2019-10-26T01:01:32.7505234Z Welcome to .NET Core!
2019-10-26T01:01:32.7506078Z ---------------------
2019-10-26T01:01:32.7506496Z Learn more about .NET Core: https://aka.ms/dotnet-docs
2019-10-26T01:01:32.7506829Z Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli-docs
2019-10-26T01:01:32.7506969Z 
2019-10-26T01:01:32.7507106Z Telemetry
2019-10-26T01:01:32.7507455Z ---------
2019-10-26T01:01:32.7507760Z The .NET Core tools collect usage data in order to help us improve your experience. The data is anonymous and doesn't include command-line arguments. The data is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.
2019-10-26T01:01:32.7508261Z 
2019-10-26T01:01:32.7508512Z Read more about .NET Core CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry
2019-10-26T01:01:33.5330617Z 
2019-10-26T01:01:33.5331729Z ASP.NET Core
2019-10-26T01:01:33.5332340Z ------------
2019-10-26T01:01:33.5332984Z Successfully installed the ASP.NET Core HTTPS Development Certificate.
2019-10-26T01:01:33.5333586Z To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only). For establishing trust on other platforms refer to the platform specific documentation.
2019-10-26T01:01:33.5335107Z For more information on configuring HTTPS see https://go.microsoft.com/fwlink/?linkid=848054.
2019-10-26T01:01:35.8326542Z Microsoft (R) Build Engine version 16.2.32702+c4012a063 for .NET Core
2019-10-26T01:01:35.8327308Z Copyright (C) Microsoft Corporation. All rights reserved.
2019-10-26T01:01:35.8327602Z 
2019-10-26T01:01:38.0153611Z   Running help generation...
2019-10-26T01:01:52.5297517Z   
2019-10-26T01:01:52.5302202Z   Target         Description
2019-10-26T01:01:52.5309166Z   ------         -----------
2019-10-26T01:01:52.5311542Z   Set-IpGroup.md No examples found
2019-10-26T01:01:52.5319772Z   
2019-10-26T01:01:52.6394678Z   A markdown file containing the help for a cmdlet is incomplete. Please check the exceptions provided for more details.
2019-10-26T01:01:52.6395199Z   At D:\a\1\s\tools\GenerateHelp.ps1:71 char:9
2019-10-26T01:01:52.6395388Z   +         throw "A markdown file containing the help for a cmdlet is in ...
2019-10-26T01:01:52.6395536Z   +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2019-10-26T01:01:52.6395688Z   + CategoryInfo          : OperationStopped: (A markdown file con\u2026d for more details.:String) [], RuntimeException
2019-10-26T01:01:52.6794108Z   + FullyQualifiedErrorId : A markdown file containing the help for a cmdlet is incomplete. Please check the exceptions provided for more details.
```

**Comment by VeryEarly** [comment]
@ssripadham
same help generation error still accoured,
it showed "Set-IpGroup.md" while your cmdlet called "SetAzApGroup",
I assume this markdown file was not needed.

Please run build/help generation/static analysis/test at local environment to verify your changes;

"msbuild build.proj /t:Build"
"msbuild build.proj /t:GenerateHelp"
"msbuild build.proj /t:StaticAnalysis"
"msbuild build.proj /t:Test"

**Comment by ssripadham** [comment]
> @ssripadham
> 
> You got following error during help generation,
> 
> ```
> Target                        Description
> Set-IpGroup.md             No examples found
> ```
> 
> Please fix the corresponding markdown file and try to verify build/help generation/static analysis/test at local to make sure CI will be passed.

What tools can I run locally to verify Markdowns?

**Comment by VeryEarly** [comment]
> > @ssripadham
> > You got following error during help generation,
> > ```
> > Target                        Description
> > Set-IpGroup.md             No examples found
> > ```
> > 
> > 
> > Please fix the corresponding markdown file and try to verify build/help generation/static analysis/test at local to make sure CI will be passed.
> 
> What tools can I run locally to verify Markdowns?

@ssripadham
Error was thrown during help generation so run 
"msbuild build.proj /t:GenerateHelp"
to make sure you fix it.

other StaticAnalysis/Test are also good to have to avoid spending time on CI.

**Comment by ssripadham** [comment]
> > > @ssripadham
> > > You got following error during help generation,
> > > ```
> > > Target                        Description
> > > Set-IpGroup.md             No examples found
> > > ```
> > > 
> > > 
> > > Please fix the corresponding markdown file and try to verify build/help generation/static analysis/test at local to make sure CI will be passed.
> > 
> > 
> > What tools can I run locally to verify Markdowns?
> 
> @ssripadham
> Error was thrown during help generation so run
> "msbuild build.proj /t:GenerateHelp"
> to make sure you fix it.
> 
I got the command to succeed locally but CI is still failing. Can you please take a look?

> other StaticAnalysis/Test are also good to have to avoid spending time on CI.
How to run these?

**Comment by ssripadham** [comment]
> > > > @ssripadham
> > > > You got following error during help generation,
> > > > ```
> > > > Target                        Description
> > > > Set-IpGroup.md             No examples found
> > > > ```
> > > > 
> > > > 
> > > > Please fix the corresponding markdown file and try to verify build/help generation/static analysis/test at local to make sure CI will be passed.
> > > 
> > > 
> > > What tools can I run locally to verify Markdowns?
> > 
> > 
> > @ssripadham
> > Error was thrown during help generation so run
> > "msbuild build.proj /t:GenerateHelp"
> > to make sure you fix it.
> 
> I got the command to succeed locally but CI is still failing. Can you please take a look?
> 
> > other StaticAnalysis/Test are also good to have to avoid spending time on CI.
> > How to run these?

dotnet.exe msbuild build.proj /t:StaticAnalysis /p:Configuration=Debug;
Microsoft (R) Build Engine version 16.2.32702+c4012a063 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Running static analysis...
  No executable found matching command "dotnet-F:\ws\Networking\azure-powershell\artifacts/StaticAnalysis/StaticAnalysis.Netcore.dll"
F:\ws\Networking\azure-powershell\build.proj(205,5): error MSB3073: The command "dotnet F:\ws\Networking\azure-powershell\artifacts/StaticAnalysis/StaticAnalysis.Netcore.dll -p F:\ws\Networking\azure-powershell\artifacts/Debug -r F:\ws\Networking\azure-powershell\artifacts//StaticAnalysisResults -s -u -m " exited with code 1.
F:\ws\Networking\azure-powershell\build.proj(211,5): error : StaticAnalysis has failed. Please follow the instructions on this doc: https://github.com/Azure/azure-powershell/blob/master/documentation/Debugging-StaticAnalysis-Errors.md

**Comment by VeryEarly** [comment]
> > > > > @ssripadham
> > > > > You got following error during help generation,
> > > > > ```
> > > > > Target                        Description
> > > > > Set-IpGroup.md             No examples found
> > > > > ```
> > > > > 
> > > > > 
> > > > > Please fix the corresponding markdown file and try to verify build/help generation/static analysis/test at local to make sure CI will be passed.
> > > > 
> > > > 
> > > > What tools can I run locally to verify Markdowns?
> > > 
> > > 
> > > @ssripadham
> > > Error was thrown during help generation so run
> > > "msbuild build.proj /t:GenerateHelp"
> > > to make sure you fix it.
> > 
> > 
> > I got the command to succeed locally but CI is still failing. Can you please take a look?
> > > other StaticAnalysis/Test are also good to have to avoid spending time on CI.
> > > How to run these?
> 
> dotnet.exe msbuild build.proj /t:StaticAnalysis /p:Configuration=Debug;
> Microsoft (R) Build Engine version 16.2.32702+c4012a063 for .NET Core
> Copyright (C) Microsoft Corporation. All rights reserved.
> 
> Running static analysis...
> No executable found matching command "dotnet-F:\ws\Networking\azure-powershell\artifacts/StaticAnalysis/StaticAnalysis.Netcore.dll"
> F:\ws\Networking\azure-powershell\build.proj(205,5): error MSB3073: The command "dotnet F:\ws\Networking\azure-powershell\artifacts/StaticAnalysis/StaticAnalysis.Netcore.dll -p F:\ws\Networking\azure-powershell\artifacts/Debug -r F:\ws\Networking\azure-powershell\artifacts//StaticAnalysisResults -s -u -m " exited with code 1.
> F:\ws\Networking\azure-powershell\build.proj(211,5): error : StaticAnalysis has failed. Please follow the instructions on this doc: https://github.com/Azure/azure-powershell/blob/master/documentation/Debugging-StaticAnalysis-Errors.md

Hi ssripadham
please go to the below link, find 
``` text
"analyze-windows/StaticAnalysisResults/SignatureIssues.csv"
```
copy content and suppress error messages into 
``` text
"azure-powershell/tools/StaticAnalysis/Exceptions/Az.Network/SignatureIssues.csv"
```

https://dev.azure.com/azure-sdk/public/_build/results?buildId=170331&view=artifacts

**Pull request merged**